packages/frontend/src/state/squeethPool/hooks.ts

```diff
@@ -0,0 +1,534 @@
+
+    await handleTransaction(
+      swapRouterContract?.methods.exactOutputSingle(exactOutputParam).send({
+        from: address,
+      }),
+    )
+  }
+  return buy
+}
+
+export const useGetBuyParamForETH = () => {
+  const address = useAtomValue(addressAtom)
+  const squeethToken = useAtomValue(squeethTokenAtom)
+  const wethToken = useAtomValue(wethTokenAtom)
+  const getBuyQuoteForETH = useGetBuyQuoteForETH()
+  const getBuyParamForETH = async (amount: BigNumber) => {
+    const quote = await getBuyQuoteForETH(amount)
+
+    return {
+      tokenIn: wethToken?.address,
+      tokenOut: squeethToken?.address,
+      fee: UNI_POOL_FEES,
+      recipient: address,
+      deadline: Math.floor(Date.now() / 1000 + 86400), // uint256
+      amountIn: ethers.utils.parseEther(amount.toString()),
+      amountOutMinimum: fromTokenAmount(quote.minimumAmountOut, OSQUEETH_DECIMALS).toString(),
+      sqrtPriceLimitX96: 0,
+    }
+  }
+
+  return getBuyParamForETH
+}
+
+export const useBuyForWETH = () => {
+  const address = useAtomValue(addressAtom)
+  const handleTransaction = useHandleTransaction()
+  const swapRouterContract = useAtomValue(swapRouterContractAtom)
+  const getBuyParamForETH = useGetBuyParamForETH()
+  const buyForWETH = async (amount: BigNumber) => {
+    const exactInputParam = await getBuyParamForETH(new BigNumber(amount))
+
+    const txHash = await handleTransaction(
+      swapRouterContract?.methods.exactInputSingle(exactInputParam).send({
+        from: address,
+        value: ethers.utils.parseEther(amount.toString()),
+      }),
+    )
+
+    return txHash
+  }
+
+  return buyForWETH
+}
+
+export const useBuyAndRefundData = () => {
+  const address = useAtomValue(addressAtom)
+  const web3 = useAtomValue(web3Atom)
+  const getBuyParamForETH = useGetBuyParamForETH()
+  const buyAndRefundData = async (amount: BigNumber) => {
+    if (!web3) return
+    const exactInputParam = await getBuyParamForETH(amount)
+    exactInputParam.recipient = address
+    const tupleInput = Object.values(exactInputParam).map((v) => v?.toString() || '')
+
+    const swapIface = new ethers.utils.Interface(routerABI)
+    const encodedSwapCall = swapIface.encodeFunctionData('exactInputSingle', [tupleInput])
+    const encodedRefundCall = swapIface.encodeFunctionData('refundETH')
+
+    return [encodedSwapCall, encodedRefundCall]
+  }
+
+  return buyAndRefundData
+}
+
+export const useBuyAndRefund = () => {
+  const address = useAtomValue(addressAtom)
+  const handleTransaction = useHandleTransaction()
+  const swapRouterContract = useAtomValue(swapRouterContractAtom)
+  const buyAndRefundData = useBuyAndRefundData()
+
+  const buyAndRefund = async (amount: BigNumber, onTxConfirmed?: () => void) => {
+    const callData = await buyAndRefundData(amount)
+
+    const result = await handleTransaction(
+      swapRouterContract?.methods.multicall(callData).send({
+        from: address,
+        value: ethers.utils.parseEther(amount.toString()),
```

Comment:
![image](https://user-images.githubusercontent.com/26922518/159376864-234686a9-ffcb-4cc7-9df0-59e463ee581c.png)
I'll get this error when I ran it locally, I need to change it to  
```suggestion
        value: fromTokenAmount(amount, 18).toString(),
```
to send the buy osqth tx

Reply:
Noticed that we used ```ethers.utils.parseEther(amount.toString())``` in a lot of places, not sure if there will be same errors

Reply:
Weird, I tried in preview and is working for me

Reply:
Yh, preview link is also working for me, but locally isn't

Reply:
@yichiehliu Not happening to me in local as well. But I noticed a weird issue. Long buy input is not working for the first time, if I click the close tab and open it again it is working. @EstherSeyi can you look into it

Reply:
okay will look into these